### PR TITLE
chore: implement an allow list on market for the sell side

### PR DESCRIPTION
**Pull request by voltaireChocolatine**
Hi vega team, we've implemented a small feature to only allow a limited amount of key to be allowed to place order on the sell side of the book. This will be useful in some future markets that we plan to implement on nebula.

I realise we could keep this feature on our fork, but it would make life easier for our validators and for future maintenance if this was added in the main repo. 

Happy to discuss further the reasons for the changes.

### CODE REVIEW

File: datanode/gateway/graphql/schema.graphql [hunk 1]
```diff
@@ -2857,6 +2857,9 @@ type Market {
 
   "Number of allowed price levels between an AMM's fair price and its quote prices. An AMM definition that exceeds this will be rejected at submission"
   allowedEmptyAMMLevels: Int!
+
+  "A list of keys allowed to plays sell orders on the market"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:2861]:
```suggestion
  "A list of keys allowed to place sell orders on the market"
```

File: datanode/gateway/graphql/schema.graphql [hunk 2]
```diff
@@ -4517,6 +4522,9 @@ enum OrderRejectionReason {
 
   "Order price exceeds the max price of the capped future market"
   ORDER_ERROR_PRICE_MUST_BE_LESS_THAN_OR_EQUAL_TO_MAX_PRICE
+
+  "ell order not allowed on this market"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4526]:
```suggestion
  "Sell order not allowed on this market"
```

File: datanode/gateway/graphql/schema.graphql [hunk 3]
```diff
@@ -4983,8 +4991,12 @@ type NewMarket {
   tickSize: String!
   "If enabled aggressive orders sent to the market will be delayed by the configured number of blocks"
   enableTxReordering: Boolean!
+
   "Number of allowed price levels between an AMM's fair price and its quote prices. An AMM definition that exceeds this will be rejected at submission"
   allowedEmptyAMMLevels: Int
+
+  "A list of keys allowed to plays sell orders on the market"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:4998]:
```suggestion
  "A list of keys allowed to place sell orders on the market"
```

File: datanode/gateway/graphql/schema.graphql [hunk 4]
```diff
@@ -5050,8 +5062,12 @@ type UpdateMarketConfiguration {
...
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5069]:
```suggestion
  "A list of keys allowed to place sell orders on the market"
```

File: datanode/gateway/graphql/schema.graphql [hunk 5]
```diff
@@ -5311,6 +5327,8 @@ type NewSpotMarket {
   tickSize: String!
   "If enabled aggressive orders sent to the market will be delayed by the configured number of blocks"
   enableTxReordering: Boolean!
+  "A list of keys allowed to plays sell orders on the market"
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5330]:
```suggestion
  "A list of keys allowed to place sell orders on the market"
```

File: datanode/gateway/graphql/schema.graphql [hunk 6]
```diff
@@ -5347,6 +5365,8 @@ type UpdateSpotMarketConfiguration {
...
```
Comment (candida-d) [inline datanode/gateway/graphql/schema.graphql:5368]:
```suggestion
  "A list of keys allowed to place sell orders on the market"
```

File: protos/sources/vega/vega.proto [hunk 7]
```diff
@@ -100,6 +100,8 @@ message StopOrder {
     REJECTION_REASON_STOP_ORDER_CANNOT_MATCH_OCO_EXPIRY_TIMES = 9;
     // Stop orders with a position size override are not supported for spot markets
     REJECTION_REASON_STOP_ORDER_SIZE_OVERRIDE_UNSUPPORTED_FOR_SPOT = 10;
+    // Stop order containing sell orders submission are not allowed for this market
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:103]:
```suggestion
    // Stop orders containing sell order submissions are not allowed for this market
```

File: protos/sources/vega/vega.proto [hunk 8]
```diff
@@ -542,6 +544,8 @@ enum OrderError {
   ORDER_ERROR_PRICE_NOT_IN_TICK_SIZE = 53;
   // Order price exceeds the max price of the capped future market
   ORDER_ERROR_PRICE_MUST_BE_LESS_THAN_OR_EQUAL_TO_MAX_PRICE = 54;
+  // Sell orders are not allowed in this market yet
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:547]:
```suggestion
  // Sell orders are not allowed in this market
```

**Review by candida-d — COMMENTED** [review]
@voltaireChocolatine I’ll leave the feature review to someone else but I’ve corrected some doc strings for the APIs

**Comment by wwestgarth** [comment]
Two thing:
- have you considered AMM's, and that someone could create an AMM to be able to indirectly sell?
- is it really the intention that people will not even be able to close out their long positions ever? Or is it that you actually only want to stop a party having a short, negative position?

**Comment by voltaireChocolatine** [comment]
>     * have you considered AMM's, and that someone could create an AMM to be able to indirectly sell?
No, but I will remove the implementation of this feature for future markets as we do not plan to use it with these, and thinking about it, it doesn't makes sense at all in the context of future markets (I believe AMM as well are only available with future markets?).

>     * is it really the intention that people will not even be able to close out their long positions ever? Or is it that you actually only want to stop a party having a short, negative position?
I think base on my previous sentence, this is being answered.

>     * this PR does not contain a single unit-test or feature test, are you sure it works?
Correct, we are still working on the tests, I've just submitted it for an early review by the vega team.

**Comment by voltaireChocolatine** [comment]
@candida-d I've merged all your changes. 
@wwestgarth I've added tests for the feature.

File: core/execution/future/market_snapshot.go [hunk 9]
```diff
@@ -220,6 +220,12 @@ func NewMarketFromSnapshot(
 	marketType := mkt.MarketType()
 
 	markPriceCalculator := common.NewCompositePriceCalculatorFromSnapshot(ctx, em.CurrentMarkPrice, timeService, oracleEngine, em.MarkPriceCalculator)
+
+	allowedSellers := map[string]struct{}{}
```
Comment (daniel1302) [inline core/execution/future/market_snapshot.go:224]: Looks like the `allowedSellers` variable is not used anywhere later.

Comment (voltaireChocolatine) [reply core/execution/future/market_snapshot.go:224]: good point, it's a leftover of a previous version which used to implement the feature on both spots and futures. I'll remove it.

**Review by EVODelavega — APPROVED** [review]

**Pull request merged**